tcms/core/middleware.py

```diff
@@ -5,15 +5,31 @@
 from django.contrib.sites.models import Site
 from django.db import DEFAULT_DB_ALIAS, connections
 from django.db.migrations.executor import MigrationExecutor
+from django.db.utils import OperationalError, ProgrammingError
+from django.http import HttpResponseRedirect
 from django.urls import reverse
 from django.utils.deprecation import MiddlewareMixin
 from django.utils.safestring import mark_safe
 from django.utils.translation import gettext_lazy as _
 
 
+class CheckDBStructureExistsMiddleware(MiddlewareMixin):
```

Comment:
This class needs to be enabled in settings.MIDDLEWARE.

However when I do enable it and it tries to redirect to /setup/ I get `no such table: django_session` because it doesn't exist and displaying the view probably relies on a few more internals than that.

IMO we can't reliably execute this from within Django so it must be a stand-alone CGI script.

Reply:
Sorry, I forgot to add the settings file to the commit. It works on my system for the three databases - sqlite, MariaDB and PostgreSQL. I'll commit it now.